File: src/main/java/dev/toastcie/mathtoastools/vector/Vector3.java
package dev.toastcie.mathtoastools.vector;

public class Vector3 {
    //TODO auto edit xy yz xz variables when x y or z is edited
    public float x;
    public float y;
    public float z;

    public Vector3(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3(Vector2 xy, float z) {
        this.x = xy.x;
        this.y = xy.y;
        this.z = z;
    }

    //editing methods

    public void normalize() {
        float length = this.length();
        this.x /= length;
        this.y /= length;
        this.z /= length;
    }

    //calculations methods

    public Vector3 add(Vector3 other) {
        return new Vector3(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    public Vector3 subtract(Vector3 other) {
        return new Vector3(this.x - other.x, this.y - other.y, this.z - other.z);
    }

    public Vector3 prod(float scalar) {
        return new Vector3(this.x * scalar, this.y * scalar, this.z * scalar);
    }

    public Vector3 cross(Vector3 other) {
        return new Vector3(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z, this.x * other.y - this.y * other.x);
    }

    public float dot(Vector3 other) {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }

    public float mix(Vector3 other, float t) {
        return this.x * (1 - t) + other.x * t + this.y * (1 - t) + other.y * t + this.z * (1 - t) + other.z * t;
    }

    //data methods

    public float length() {
        return (float) Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }
}

File: src/main/kotlin/dev/toastcie/mathtoastools/vector/Vector3.kt
package dev.toastcie.mathtoastools.vector

import kotlin.math.sqrt

/**
 * A 3D vector
 * @property x the x value of the vector
 * @property y the y value of the vector
 * @property z the z value of the vector
 * @constructor creates a vector with the given values
 */
data class Vector3(var x: Float, var y: Float, var z: Float) {
    //constructors
    constructor(xy: Vector2, z: Float) : this(xy.x,xy.y,z)

    //editing methods
    /**
     * the normalized vector
     */
    val normalize: Vector3
        get(){
            val len = length
            return Vector3(x / len, y / len, z / len)
        }

    //specials vectors
    /**
     * the x and y values of the vector
     */
    val xy: Vector2
        get() = Vector2(x, y)

    /**
     * the x and z values of the vector
     */
    val xz: Vector2
        get() = Vector2(x, z)

    /**
     * the y and z values of the vector
     */
    val yz: Vector2
        get() = Vector2(y, z)

    //calculations methods
    /**
     * Adds the x and y values of the vector
     * @param other the vector to add
     * @return the new vector
     */
    operator fun plus(other: Vector3): Vector3 = Vector3(x + other.x, y + other.y, z + other.z)
    /**
     * Adds the x and y values of the vector
     * @param other the float to add
     * @return the new vector
     */
    operator fun plus(other: Float): Vector3 = Vector3(x + other, y + other, z + other)

    /**
     * Subtracts the x and y values of the vector
     * @param other the vector to subtract
     * @return the new vector
     */
    operator fun minus(other: Vector3): Vector3 = Vector3(x - other.x, y - other.y, z - other.z)
    /**
     * Subtracts the x and y values of the vector
     * @param other the float to subtract
     * @return the new vector
     */
    operator fun minus(other: Float): Vector3 = Vector3(x - other, y - other, z - other)

    /**
     * multiply the x and y values of the vector
     * @param other the vector to multiply
     * @return the new vector
     */
    operator fun times(other: Vector3): Vector3 = Vector3(
        y * other.z - z * other.y,
        z * other.x - x * other.z,
        x * other.y - y * other.x
    )
    /**
     * multiply the x and y values of the vector
     * @param other the float to multiply
     * @return the new vector
     */
    operator fun times (scalar: Float): Vector3 = Vector3(x * scalar, y * scalar, z * scalar)

    /**
     * dot product of 2 vectors
     * @param other the vector to dot
     * @return the dot product
     */
    infix fun dot(other: Vector3): Float = x * other.x + (y * other.y) + (z * other.z)

    /**
     * mix 2 vectors
     * @param other the vector to mix
     * @param t the mix value
     * @return the mixed vector
     */
    fun mix(other: Vector3, t: Float): Float = x * (1 - t) + (other.x * t) + (y * (1 - t)) + (other.y * t) + (z * (1 - t)) + (other.z * t)

    //data methods
    /**
     * the length of the vector
     */
    val length: Float
            get() = sqrt((x * x + (y * y) + (z * z)))
}